src/workflows/airqo_etl_utils/bigquery_api.py

```
@@ -1384,7 +1366,83 @@ def generate_missing_data_query(
+            )
+            -- Find missing data by comparing expected vs actual
+            SELECT
+                edp.device_id,
+                edp.timestamp
+            FROM expected_data_points edp
+            LEFT JOIN actual_data ad
+                ON edp.device_id = ad.device_id
+                AND edp.timestamp = ad.timestamp
+            WHERE ad.device_id IS NULL  -- Only return where there's no match (missing data)
+            ORDER BY edp.device_id, edp.timestamp;
+        """
```

Ruff (0.14.1)

Possible SQL injection vector through string-based query construction

(S608)